### PR TITLE
I166 don't send device register call multiple times

### LINKED ISSUE #166: Register mobile device for push notification after login
## Feature Request
Coming in alignment with #160, this issue addresses the mobile application related changes necessary to implement this. 

**Is your feature request related to a problem? Please describe:**
A clear and concise description of the problem or need that the feature would address.

**Proposed Solution:**
After successful login, if the device is not registered, the device will obtain/generate a unique device token that will be sent to the `/register/${userId}` endpoint for registration. Upon successful registration, it will keep the information that the device has been registered successfully so that it does not sent the register request every time the app opens or user logins.

## Acceptance Criteria

Please describe the minimum acceptance criteria or requirements that need to be met for this feature to be considered complete.

- [x] After login, if the device is not registered to the backend for push notifications, it will make a POST request to the  `/register/${userId}` endpoint for registering the device for push notification.
- [x] After successful registration, it will store the information locally so that it does not register to the backend multiple times.
- [ ] If already registered, it will not send the register request.

### PR DESCRIPTION
**Pull request by Rezwanul-Haque**
Fixes #166 

Resolves #
Used when the changes in the pull request indirectly resolve another issue.

- [ ] Does not depend on other PRs
- [x] Depends on other PR being merged first: [#239](https://github.com/CraftsmenLtd/BloodConnect/pull/239)

## What kind of change does this PR introduce (check at least one)

- [ ] Bugfix
- [ ] Improvement
- [x] New feature
- [x] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?
Device registration will be ignored if the user already registered with the same device token.

Provide a brief overview of the changes introduced in this pull request.
Previously on every login we try to register the device, with this implementation we are ignoring the registration if the user 
already registered with the same device token.

## Related Issues/PRs (optional)

Link to the relevant issue(s) or provide a clear description of the problem or feature being addressed.

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [x] I have tested these changes locally and they function as expected.
- [x] I have added or updated necessary documentation to reflect the changes.
- [x] I have followed the project's coding style and guidelines.
- [ ] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [ ] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Reviewer Request (optional)

If you have any specific requests or suggestions for reviewers, mention them here.

## Screenshots (optional)

If applicable, provide any additional screenshots or GIFs that showcase the changes made in this pull request.

## Demo Link (optional)

If applicable, provide a link to a live demo or deployment where the changes can be reviewed.

## Known Issues (optional)

If there are any known issues or limitations with the changes made in this pull request, describe them here.

## Additional Notes (optional)

Add any additional notes or comments you think are relevant to this pull request.

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [x] No risk
- [ ] Low
- [ ] High

### CODE REVIEW

**Review by EbrahimAli-RU — CHANGES_REQUESTED** [review]

File: clients/mobile/src/setup/notification/saveDeviceToken.ts [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
+import StorageService from '../../utility/storageService'
+import { TOKEN } from '../constant/token'
 
 export const saveDeviceTokenToSNS = async(
   deviceToken: string,
   fetchClient: any
 ): Promise<void> => {
   try {
+    if (await checkDeviceAlreadyRegisteredForSameUser()) {
+      console.log('Device already registered with the same user, skipping registration again')
```
Comment (sakibstark11) [inline clients/mobile/src/setup/notification/saveDeviceToken.ts:12]: are you sure we're using console.logs? or should we use pino? the lambda stuff I belive uses pino

File: clients/mobile/src/setup/notification/saveDeviceToken.ts [hunk 2]
```diff
@@ -1,10 +1,18 @@
 import { Platform } from 'react-native'
+import authService from '../../authentication/services/authService'
+import StorageService from '../../utility/storageService'
+import { TOKEN } from '../constant/token'
 
 export const saveDeviceTokenToSNS = async(
```
Comment (sakibstark11) [inline clients/mobile/src/setup/notification/saveDeviceToken.ts:6]: the name of this wrong

saveDeviceTokenOnSNS

File: clients/mobile/src/setup/notification/saveDeviceToken.ts (hunk 1 above)
Comment (Biplob14) [reply clients/mobile/src/setup/notification/saveDeviceToken.ts:12]: This is for mobile application, `console.log` will be shown only on terminal for debugging and there is no other purpose of this in real Application.

File: clients/mobile/src/setup/notification/saveDeviceToken.ts [hunk 3]
```diff
@@ -24,3 +32,23 @@ export const saveDeviceTokenToSNS = async(
...
+  deviceToken: string
+): Promise<void> => {
+  const loggedInUser = await authService.currentLogInUser()
+  await StorageService.storeItem<{ deviceToken: string; userId: string }>(
+    TOKEN.DEVICE_TOKEN,
+    { userId: loggedInUser.userId, deviceToken }
+  )
+}
+
+export const checkDeviceAlreadyRegisteredForSameUser = async(): Promise<boolean> => {
```
Comment (sakibstark11) [inline clients/mobile/src/setup/notification/saveDeviceToken.ts:46]: this name is also a bit strange
`isDeviceAlreadyRegisteredForUser`
Maybe it should be passed current `loggedInUser.userId`, to make it simple.

Then, you can do something like


```
if (!(await isDeviceAlreadyRegisteredForUser(deviceId, userId))) {
// do something
return
}
return 
```

**Comment by Biplob14** [comment]
Please follow out commit convention and edit your commit message.

File: clients/mobile/src/authentication/services/authService.ts [hunk 4]
```diff
@@ -112,6 +133,20 @@ export const fetchSession = async(): Promise<FetchSessionResponse> => {
...
+  if (session?.tokens === undefined) {
+    throw new Error('Session or tokens are undefined')
+  }
+
+  return {
+    email: session.tokens.idToken?.payload.email as string,
+    userId: session.tokens.idToken?.payload['custom:userId'] as string,
+    phoneNumber: session.tokens.idToken?.payload.phone_number as string,
+    name: session.tokens.idToken?.payload.name as string
+  } satisfies User
```
[the remark this answers is not in the document]
Comment (Rezwanul-Haque) [reply clients/mobile/src/authentication/services/authService.ts:147]: `satisfies User` ensures that the returned object matches the structure defined by the User interface. If the object is missing properties or has incorrect types for any property, TypeScript will throw an error at compile time.

File: clients/mobile/src/setup/notification/saveDeviceToken.ts (hunk 2 above)
Comment (Rezwanul-Haque) [reply clients/mobile/src/setup/notification/saveDeviceToken.ts:6]: I didn't touch the existing function name. will change.

File: clients/mobile/src/authentication/services/authService.ts [hunk 5]
```diff
@@ -194,5 +229,6 @@ export default {
   loginUser,
   logoutUser,
   fetchSession,
-  decodeAccessToken
+  decodeAccessToken,
+  currentLogInUser: currentLoggedInUser
```
[the remark this answers is not in the document]
Comment (Rezwanul-Haque) [reply clients/mobile/src/authentication/services/authService.ts:233]: it's changed by my IDE when renaming the function. My mistake didn't noticed.

File: clients/mobile/src/setup/notification/saveDeviceToken.ts (hunk 2 above)
Comment (Rezwanul-Haque) [reply clients/mobile/src/setup/notification/saveDeviceToken.ts]: resolved in [9a0e8d2a6](https://github.com/CraftsmenLtd/BloodConnect/pull/272/commits/9a0e8d2a659996bbf0a9183fdf1252e7f074ca9f)

File: clients/mobile/src/setup/notification/saveDeviceToken.ts (hunk 3 above)
Comment (Rezwanul-Haque) [reply clients/mobile/src/setup/notification/saveDeviceToken.ts]: resolved in [9a0e8d2a6](https://github.com/CraftsmenLtd/BloodConnect/pull/272/commits/9a0e8d2a659996bbf0a9183fdf1252e7f074ca9f)

File: clients/mobile/src/authentication/services/authService.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rezwanul-Haque) [reply clients/mobile/src/authentication/services/authService.ts]: resolved in [7c20661b8](https://github.com/CraftsmenLtd/BloodConnect/pull/272/files#diff-7c20661b8328c2f34f993685ba0e900c2584db81b3c59c52bfcf71800295ac49R233)

File: clients/mobile/src/setup/notification/saveDeviceToken.ts (hunk 1 above)
Comment (Rezwanul-Haque) [reply clients/mobile/src/setup/notification/saveDeviceToken.ts]: removed

File: clients/mobile/src/userWorkflow/Profile.tsx [hunk 6]
```diff
@@ -64,17 +64,18 @@ const posts: Post[] = [
   const theme = useTheme()
   const auth = useAuth()
 
-  const clearAllStorage = async () => {
+  const clearStorageExceptDeviceToken = async () => {
     if (Platform.OS !== 'web') {
-      await AsyncStorage.clear();
+      const allKeys = await AsyncStorage.getAllKeys();
+      const keysToRemove = allKeys.filter(key => key !== TOKEN.DEVICE_TOKEN);
+      await AsyncStorage.multiRemove(keysToRemove);
```
[the remark this answers is not in the document]
Comment (Rezwanul-Haque) [reply clients/mobile/src/userWorkflow/Profile.tsx]: resolved in [e1879f2864a](https://github.com/CraftsmenLtd/BloodConnect/pull/272/files#diff-e1879f2864af3cc95c9f1bc23dbd66a0329bbb5f2cb2dac777bd77569ed281f2R79-R81)

File: clients/mobile/.env.example [hunk 7]
```diff
@@ -0,0 +1,13 @@
+COUNTRY=
```
Comment (sakibstark11) [inline clients/mobile/.env.example:1]: don't leave things empty
give examples of the kind and format of values that needs setting

File: clients/mobile/src/utility/storageService.ts [hunk 8]
```diff
@@ -41,9 +41,87 @@ const clearStorage = async(): Promise<void> => {
...
+  getKeys(): readonly string[];
+
+  /**
+   * Removes all keys in the set from storage.
+   * @returns A promise that resolves when the keys have been removed.
+   */
+  remove(): Promise<void>;
+}
+
+/**
```
Comment (sakibstark11) [inline clients/mobile/src/utility/storageService.ts:68]: why if we already have typescript

**Review by EbrahimAli-RU — APPROVED** [review]

File: clients/mobile/src/utility/storageService.ts (hunk 8 above)
Comment (EbrahimAli-RU) [reply clients/mobile/src/utility/storageService.ts:68]: Vai, The code looks messy and hard to read.

Comment (EbrahimAli-RU) [reply clients/mobile/src/utility/storageService.ts:68]: Vai, The code looks messy and hard to read.

**Review by EbrahimAli-RU — CHANGES_REQUESTED** [review]

File: clients/mobile/src/utility/storageService.ts (hunk 8 above)
Comment (Rezwanul-Haque) [reply clients/mobile/src/utility/storageService.ts]: removed

File: .gitignore [hunk 9]
```diff
@@ -1,5 +1,6 @@
 # IDE
 .idea
+*~
```
Comment (sakibstark11) [inline .gitignore:3]: what is this

Comment (Rezwanul-Haque) [reply .gitignore:3]: Bhai my IDE was generating multiple backup files(a settings enabled) with that signature(storageService.ts~) whenever I made a change in any file and when committing those files also included, So with this I'm ignoring those files.

**Review by Biplob14 — APPROVED** [review]

**Review by EbrahimAli-RU — APPROVED** [review]

**Pull request merged**